service/qos/service_level_controller.cc

```diff
@@ -220,6 +220,72 @@ future<> service_level_controller::update_service_levels_cache() {
     });
 }
 
+static std::unordered_set<sstring> get_all_roles(const std::unordered_multimap<sstring, sstring>& hierarchy, const sstring& base_role) {
+    std::unordered_set<sstring> roles;
+    std::queue<sstring> queue;
+    queue.push(base_role);
+
+    while (!queue.empty()) {
+        auto role = queue.front();
+        auto [it, it_end] = hierarchy.equal_range(role);
+
+        while (it != it_end) {
+            if (!roles.contains(it->second)) {
+                queue.push(it->second);
+            }
+            ++it;
+        }
+        roles.insert(role);
+        queue.pop();
+    }
+
+    return roles;
+}
+
+future<> service_level_controller::update_effective_service_levels_cache() {
+    assert(this_shard_id() == global_controller);
+
+    return with_semaphore(_global_controller_db->notifications_serializer, 1, [this] () {
+        return async([this] () {
+            auto& role_manager = _auth_service.local().underlying_role_manager();
+            const auto hierarchy = role_manager.query_all_directly_granted().get();
+            // includes only roles with attached service level
+            const auto attributes = role_manager.query_attribute_for_all("service_level").get();
+
+            std::map<sstring, service_level_options> effective_sl_map;
+            ::parallel_for_each(attributes.begin(), attributes.end(), [&hierarchy, &attributes, &effective_sl_map, this] (auto& attribute_entry) -> future<> {
+                auto& base_role = attribute_entry.first;
+
+                auto all_roles = get_all_roles(hierarchy, base_role);
```

Comment:
It looks like you are going over all roles, for each role you compute all their "descendants" (not sure if that's the proper term here), then you combine them to calculate the effective service level. Is this necessary to always enumerate all descendant and combine?

I know that it's more efficient now that we have everything in memory, but it looks like we could benefit from some memoization: it looks to me that when enumerating service levels, if you encounter a service level for which you already computed the effective service level, you could take that effective service level and not have to push the whole subtree.

IIUC combining service levels is an associative and commutative operation, if so then this should work and should give you O(N) running time for refreshing all service levels.